android-agent/src/main/java/io/opentelemetry/android/internal/services/ServiceManager.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.android.internal.services
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import com.squareup.wire.internal.toUnmodifiableMap
+import io.opentelemetry.android.internal.services.periodicwork.PeriodicWorkService
+
+/**
+ * This class is internal and not for public use. Its APIs are unstable and can change at any time.
+ */
+class ServiceManager
```

Comment:
Another alternative to organize this is to expose a public interface for `ServiceManager` that contains all the methods folks will call (e.g. `getPreferencesService()`), make the implementation `internal`, and then just create an instance of the object and mark that as `@JvmStatic` so that the singleton is accessible.

That way, you don't need to change the visible of the constructor or mark as being visible for testing, or have methods that allows you to replace the instance in the companion object just for testing - you can just create a new instance in tests instead of relying on the singleton used at runtime. To do that, you can pass in an instance of this in via the constructor of objects that depend on it instead of referencing the singleton directly. You can even create a `FakeServiceManager` for testing objects that require this, but the tests don't actually need a real instance.

This way of organizing this will reduce the amount of test-only scaffolding you need, and generally make your tests easier to write.

Reply:
Sounds good! I'll take a look into this approach 👍 thanks.

Reply:
I think it should now match what you've described @bidetofevil